### PR TITLE
Accidentally removed CA should trigger crt recreation (and rolling update)

### PR DESCRIPTION
**Pull request by stanlyDoge**
Signed-off-by: Stanislav Knot <sknot@redhat.com>

### Type of change
- Bugfix

### Description
Fixes https://github.com/strimzi/strimzi-kafka-operator/issues/542

### Checklist
- [ ] Update/write design documentation in `./design`
- [x] Write tests
- [x] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

File: operator-common/src/main/java/io/strimzi/operator/cluster/model/Ca.java [hunk 1]
```diff
@@ -737,6 +741,10 @@ public boolean keyReplaced() {
         return renewalType == RenewalType.REPLACE_KEY;
     }
 
+    public boolean keyCreated() {
+        return renewalType == RenewalType.CREATE;
```
Comment (samuel-hawker) [inline operator-common/src/main/java/io/strimzi/operator/cluster/model/Ca.java]: This is mostly stylistic but I believe `.equals` is generally preferred to `==` as we can choose to override how we compare objects and can avoid object hashes not matching https://stackoverflow.com/questions/7520432/what-is-the-difference-between-and-equals-in-java
99% of the time it doesn't matter but thought it was worth a mention.

File: systemtest/src/test/java/io/strimzi/systemtest/RollingUpdateST.java [hunk 2]
```diff
@@ -777,6 +778,41 @@ void testTriggerRollingUpdateAfterOverrideBootstrap() throws CertificateExceptio
         // TODO: send and recv messages via this new bootstrap (after client builder) https://github.com/strimzi/strimzi-kafka-operator/pull/2520
     }
 
+    @Test
+    void testAccidentallyRemovedCaTriggersRollingUpdate() {
```
Comment (samuel-hawker) [inline systemtest/src/test/java/io/strimzi/systemtest/RollingUpdateST.java]: Might be overly nit-picky, but the cause of why it is removed could be seen as not part of the 'test'
I would opt to have it be called something like
```
testClusterCaRemovedTriggersRollingUpdate
```
and then have a documented comment explaining the context of this test is it being accidentally removed.

File: operator-common/src/main/java/io/strimzi/operator/cluster/model/Ca.java [hunk 3]
```diff
@@ -737,6 +741,10 @@ public boolean keyReplaced() {
         return renewalType == RenewalType.REPLACE_KEY;
     }
 
+    public boolean keyCreated() {
```
Comment (samuel-hawker) [inline operator-common/src/main/java/io/strimzi/operator/cluster/model/Ca.java]: `isKeyCreated` would be slightly more idiomatic, but given the other similar methods are not using `is` I can understand why you matched the convention

File: systemtest/src/test/java/io/strimzi/systemtest/RollingUpdateST.java [hunk 4]
```diff
@@ -777,6 +778,41 @@ void testTriggerRollingUpdateAfterOverrideBootstrap() throws CertificateExceptio
...
+        KafkaResource.kafkaPersistent(CLUSTER_NAME, 3, 3).done();
+        String topicName = "test-topic-" + new Random().nextInt(Integer.MAX_VALUE);
+        KafkaTopicResource.topic(CLUSTER_NAME, topicName, 2, 2).done();
+
+        String userName = "alice";
+        KafkaUser user = KafkaUserResource.tlsUser(CLUSTER_NAME, userName).done();
```
Comment (see-quick) [inline systemtest/src/test/java/io/strimzi/systemtest/RollingUpdateST.java]: use please username from `BaseST`  & remove that redundant definition
```suggestion
        KafkaUser user = KafkaUserResource.tlsUser(CLUSTER_NAME, USER_NAME).done();
```

File: systemtest/src/test/java/io/strimzi/systemtest/RollingUpdateST.java [hunk 5]
```diff
@@ -777,6 +778,41 @@ void testTriggerRollingUpdateAfterOverrideBootstrap() throws CertificateExceptio
...
+
+        Map<String, String> kafkaPods = StatefulSetUtils.ssSnapshot(KafkaResources.kafkaStatefulSetName(CLUSTER_NAME));
+        Map<String, String> zkPods = StatefulSetUtils.ssSnapshot(KafkaResources.zookeeperStatefulSetName(CLUSTER_NAME));
+
+        int sent = internalKafkaClient.sendMessagesTls(topicName, NAMESPACE, CLUSTER_NAME, userName, MESSAGE_COUNT, "TLS");
+        assertThat(sent, is(MESSAGE_COUNT));
+
+        String zkCrtBeforeAccident = kubeClient(NAMESPACE).getSecret(CLUSTER_NAME + "-zookeeper-nodes").getData().get(CLUSTER_NAME + "-zookeeper-0.crt");
+        String kCrtBeforeAccident = kubeClient(NAMESPACE).getSecret(CLUSTER_NAME + "-kafka-brokers").getData().get(CLUSTER_NAME + "-kafka-0.crt");
+        kubeClient(NAMESPACE).deleteSecret(KafkaResources.clusterCaKeySecretName(CLUSTER_NAME));
```
Comment (see-quick) [inline systemtest/src/test/java/io/strimzi/systemtest/RollingUpdateST.java]: Do we need to explicitly choose namespace? Will the following simply not work?
```suggestion
        kubeClient().deleteSecret(KafkaResources.clusterCaKeySecretName(CLUSTER_NAME));
```

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ModelUtils.java [hunk 6]
```diff
@@ -234,14 +234,14 @@ public static Secret buildSecret(ClusterCa clusterCa, Secret secret, String name
             reasons.add("certificate doesn't exist yet");
             shouldBeRegenerated = true;
         } else {
-            if (clusterCa.certRenewed() || (clusterCa.isExpiring(secret, keyCertName + ".crt") && isMaintenanceTimeWindowsSatisfied)) {
+            if (clusterCa.keyCreated() || clusterCa.certRenewed() || (clusterCa.isExpiring(secret, keyCertName + ".crt") && isMaintenanceTimeWindowsSatisfied)) {
```
[the remark this answers is not in the document]
Comment (stanlyDoge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/ModelUtils.java]: `isMaintenanceTimeWindowsSatisfied` is tijed just for this part of condition: `(clusterCa.isExpiring(secret, keyCertName + ".crt") && isMaintenanceTimeWindowsSatisfied)`
However it makes sense to move is as the first one in this scope. Good catch.

**Pull request merged**